packages/hierarchies/learning/HierarchyDefinitions.md

```diff
@@ -0,0 +1,271 @@
+                ec_classname(ECClassId, 's.c') ClassName,
+                ECInstanceId Id,
+                UserLabel Label
+              FROM
+                BisCore.PhysicalElement
+            `,
+          },
+        },
+      ];
+    }
+    // Otherwise, return an empty array to indicate that there are no children
+    return [];
+  },
+  parseNode(row) {
+    // Parse the row into an instance node
+    return {
+      key: {
+        type: "instances",
+        instanceKeys: [{ className: row.ClassName, id: row.Id }],
+      },
+      label: row.Label,
+    };
+  },
+};
+```
+
+<!-- END EXTRACTION -->
+
+## Custom pre-processing
+
+In some situations, the hierarchy definition may want to modify or remove some nodes from the hierarchy before the processing (hiding, grouping, sorting) is performed. This can be achieved by providing a custom `preProcessNode` callback.
+
+For example, the following code snippet shows how to use the `preProcessNode` callback to modify and remove nodes based on external service response:
+
+<!-- [[include: [Presentation.Hierarchies.HierarchyDefinitions.Imports, Presentation.Hierarchies.HierarchyDefinitions.PreProcessNode], ts]] -->
+<!-- BEGIN EXTRACTION -->
+
+```ts
+import { createClassBasedHierarchyDefinition, createNodesQueryClauseFactory, HierarchyDefinition, HierarchyNode } from "@itwin/presentation-hierarchies";
+
+const hierarchyDefinition: HierarchyDefinition = {
+  async defineHierarchyLevel({ parentNode }) {
+    // For root nodes, return all physical elements
+    if (!parentNode) {
+      const queryClauseFactory = createNodesQueryClauseFactory({ imodelAccess });
+      return [
+        {
+          fullClassName: "BisCore.PhysicalElement",
+          query: {
+            ecsql: `
+              SELECT ${await queryClauseFactory.createSelectClause({
+                ecClassId: { selector: "x.ECClassId" },
+                ecInstanceId: { selector: "x.ECInstanceId" },
+                nodeLabel: { selector: "x.UserLabel" },
+              })}
+              FROM BisCore.PhysicalElement x
+            `,
+          },
+        },
+      ];
+    }
+    // Otherwise, return an empty array to indicate that there are no children
+    return [];
+  },
+  async preProcessNode(node) {
+    // The pre-processor queries an external service to get an external ID for the node
+    // and either adds it to the node's extended data or omits the node from the hierarchy
+    // if the external ID is not found.
+    const externalId = await externalService.getExternalId(node);
+    if (externalId) {
+      return { ...node, extendedData: { ...node.extendedData, externalId } };
+    }
+    return undefined;
+  },
+};
+```
+
+<!-- END EXTRACTION -->
+
+## Custom post-processing
+
+The post-processing step happens after nodes' hiding, grouping and children determination steps, but before sorting. The step allows modifying the nodes and is especially useful when grouping nodes need to be modified in some way.
```

Comment:
If I were you, I'd reword the second sentence to avoid repeating modifying/modified and explain why this is useful: 

```suggestion
The post-processing step happens after nodes' hiding, grouping and children determination steps, but before sorting. This step allows to modify the nodes and is especially useful for the grouping nodes because this is the only way to customize them before completing the hierarchy level definition.
```

Reply:
What do you think about the following?

_The post-processing step allows to modify the nodes in their final form - fully customized, grouped and with children determined. Also, since the step happens after grouping, it gives a chance to apply modifications to grouping nodes._

Reply:
That's much better

Reply:
pushed